doc/neps/nep-0029-deprecation_policy.rst

```diff
@@ -0,0 +1,294 @@
+   - support minor versions of ``numpy`` initially released in the 24
+     months prior to a planned project release date or the oldest
+     version that supports the minimum Python version (whichever is
+     higher)
+
+   The minimum supported version of ``Python`` will be set to
+   ``python_requires`` in ``setup``.  All supported versions of
+   Python will be in the test matrix and have binary artifacts built
+   for releases.
+
+   The project will bump (adjust upward) the minimum Python and NumPy
+   version support on every minor and major release, but never on a
+   patch release.
+
+For other dependencies, adopt similar time windows of 24 months or
+shorter.
+
+
+Backward compatibility
+----------------------
+
+No backward compatibility issues.
+
+Alternatives
+------------
+
+Ad-Hoc version support
+~~~~~~~~~~~~~~~~~~~~~~
+
+A project could on every release evaluate whether to increase
+the minimum version of Python supported.
+As a major downside, an ad-hoc approach makes it hard for downstream users to predict what
+the future minimum versions will be.  As there is no objective threshold
+to when the minimum version should be dropped, it is easy for these
+version support discussions to devolve into [bike shedding](https://en.wikipedia.org/wiki/Wikipedia:Avoid_Parkinson%27s_bicycle-shed_effect) and acrimony.
+
+
+All CPython supported versions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The CPython supported versions of Python are listed in the Python
+Developers Guide and the Python PEPs. Supporting these are a very
+clear and conservative approach.  However, it means that there is 4
+year lag between when new language features come into the language and
+when the projects are able to use them.  Additionally, for projects
+that have a significant component of compiled extensions this requires
+building many binary artifacts for each release.
+
+For the case of NumPy, many projects carry workarounds to bugs that
+are fixed in subsequent versions of NumPy.  Being proactive about
+increasing the minimum version of NumPy will allow downstream
+packages to carry fewer version-specific patches.
+
+
+
+Default version on Linux distribution
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The policy could be to support the version of Python that ships by
+default in the latest Ubuntu LTS or CentOS/RHEL release.  However, we
+would still have to standardize across the community which
+distribution we are following.
+
+By following the versions supported by major Linux distributions, we
+are giving up technical control of our projects to external
+organizations that may have different motivations and concerns than we
+do.
+
+N minor versions of Python
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Given the current release cadence of the Python, the proposed time
+(42 months) is roughly equivalent to "the last two" Python minor
+versions.  However, if Python changes their release cadence, any rule
+based on the number of minor releases will need to be changed.
+
```

Comment:
A rule based on the number of minor releases wouldn't _inherently_ have to change if the release cadence were to change by epsilon. Only if the change in release cadence is sufficiently large might the rule might have to change, and then only if we require that the rule remain sensible : )
I think that the real point is that the rule shouldn't be based _solely_ on minor versions **or** _solely_ on time; a composite rule is more robust to changes in release cadence.

Reply:
I think it's also reasonable for the Scientific Python community to ask Python core to use "Impact on the Scientific Python community's standard support policy" as one of our points of consideration if we're considering adjusted our release cadence (which we are actually doing at the moment - we'd like to align it with the calendar year, which means either an every-year cadence, or an every-other-year one).

So if you want to consider that potential impact in advance and design a rule around it, the two scenarios to consider would be:

1. What if the current 18 months became 12 months?
2. What if the current 18 months became 24 months?

(If attempting to design such a rule results in a conclusion of "That cadence would be awful for us", then that's important feedback for us to hear *before* we make any decisions)

Reply:
> What if the current 18 months became 12 months?

This would probably be fine, and we could leave this policy unchanged at 42 months - that would make packages oscillate between supporting 3 and 4 Python releases, rather than 2-3. Taking a year off our support would be too aggressive imho - we'd get many more users who we're then forcing to upgrade their Python install (can be painful).

> What if the current 18 months became 24 months?

that would be awful

Reply:
Note that there's more to the 24 month proposal than slowing down the breaking releases - there are 12 months of backwards compatible incremental feature releases, so both proposals reduce the  "no CPython feature releases" gap to at most 12 months. It's just one of them also aims to reduce the frequency of "rebuild the world" events.